File: 01a/js/src/main.ts
import express from 'express'
import YAML from 'yaml'
import { parse } from 'csv-parse/sync';
import bodyParser from 'body-parser';

const app = express()
const port = 5173

// parse all bodies to text in req.body,
// so we can manually handle parsing afterwards
app.use(bodyParser.text({ type: ['application/yaml', 'application/json', 'text/csv', 'text/plain'] }));

app.post("/parse", (req, res) => {
  // https://mimetype.io/all-types
  // List of MIME types.
  if (req.headers['content-type'] == "text/plain") {
    let person = myNovelTextFormatParser(req.body)
    console.log(person)
    res.send(person)
  }
  if (req.headers['content-type'] == "text/csv") {
    const rows = parse(req.body, {
      columns: true,
      skip_empty_lines: true,
      delimiter: ";"
    })
    console.log(rows)
    res.send(rows)
  }
  if (req.headers['content-type'] == "application/json") {
    const json = JSON.parse(req.body)
    console.log(json)
    res.send(json)
  }
  if (req.headers['content-type'] == "application/yaml") {
    const yaml = YAML.parse(req.body)
    console.log(yaml)
    res.send(yaml)
  }
})

app.listen(port, () => {
  console.log(`Listening on port ${port}`)
})

function myNovelTextFormatParser(body: String) {
  let person = {} as personObject

  // split string at the start of every header
  let sections = body.split("[")

  sections.forEach((section) => {
    let objectName = section.substring(0, section.indexOf("]"))

    // https://stackoverflow.com/questions/21895233/how-to-split-string-with-newline-n-in-node#comment77609681_21895354
    // Covering all newline edge cases with a regex.
    let fieldsLines = section.substring(section.indexOf("]")).split(/[\r\n]+/)

    // Clean up
    fieldsLines.shift()
    if (fieldsLines[-1] === "") {
      fieldsLines.pop()
    }

    let fieldsObjects: any[] = []
    fieldsLines.forEach((field) => {
      let field_name_and_value = field.split(": ")
      fieldsObjects.push(Object.assign({}, field_name_and_value))
    })

    if (objectName == "me") {
      let me: meObject = {
        name: fieldValueForName("name", fieldsObjects),
        occupation: fieldValueForName("occupation", fieldsObjects)
      }
      person.me = me
    }

    if (objectName == "studies") {
      let studies: studiesObject = {
        name: fieldValueForName("name", fieldsObjects),
        address: fieldValueForName("address", fieldsObjects),
        program: fieldValueForName("program", fieldsObjects),
      }
      person.studies = studies
    }

    if (objectName == "hobbies") {
      let hobbies: string[] = []
      fieldsObjects.forEach((fieldObject) => {
        hobbies.push(fieldObject[0])
      })
      person.hobbies = hobbies
    }
  })

  return person
}

function fieldValueForName(name: String, fieldsObjects: String[]): String {
  let fieldValue = fieldsObjects.find((object) => object[0] == name)
  if (fieldValue != undefined) {
    fieldValue = fieldValue[1]
    return fieldValue
  } else {
    throw new Error("Illegal field name detected. Check your text for mistakes.")
  }
}

interface meObject {
  name: String
  occupation: String
}

interface studiesObject {
  name: String,
  address: String,
  program: String
}

interface personObject {
  me: meObject,
  studies: studiesObject,
  hobbies: string[]
}

File: 01a/python/main.py
import uvicorn
from fastapi import FastAPI, Request
import yaml
import pandas as pd
import io
import pprint

app = FastAPI()


@app.post("/parse")
async def parse(req: Request):
    content_type = req.headers.get("Content-Type")
    if content_type == "application/json":
        print(await req.json())
    if content_type == "application/yaml":
        print(yaml.safe_load(await req.body()))
    if content_type == "text/csv":
        dataframa = pd.read_csv(
            io.StringIO(str(await req.body(), encoding="utf-8")), sep=";"
        )
        person_dict = dataframa.to_dict()
        pprint.pprint(person_dict)
    if content_type == "text/plain":
        person = my_novel_text_format_parser(str(await req.body(), encoding="utf-8"))
        pprint.pprint(person)


def my_novel_text_format_parser(body: str):
    person = []

    sections = body.split("[")
    sections.pop(0)

    for section in sections:
        split_string = section.split("]")
        object_name = split_string[0]
        fields = split_string[1]

        fields_lines = fields.strip()
        fields_lines = fields_lines.splitlines()

        fields_dicts = []
        for field in fields_lines:
            field_name_and_value = field.split(": ")

            if len(field_name_and_value) == 1:
                field_name_and_value.append("")

            fields_dicts.append(
                dict(name=field_name_and_value[0], value=field_name_and_value[1])
            )

        if object_name == "me":
            me = dict(
                name=field_value_for_name("name", fields_dicts),
                occupation=field_value_for_name("occupation", fields_dicts),
            )
            person.append(me)

        if object_name == "studies":
            studies = dict(
                name=field_value_for_name("name", fields_dicts),
                address=field_value_for_name("address", fields_dicts),
                program=field_value_for_name("program", fields_dicts),
            )
            person.append(studies)

        if object_name == "hobbies":
            hobbies = []
            for field in fields_dicts:
                hobbies.append(list(field.values())[0])
            person.append(hobbies)

    return person


def field_value_for_name(name: str, fields_dicts: []):
    for field in fields_dicts:
        if list(field.values())[0] == name:
            return list(field.values())[1]


if __name__ == "__main__":
    uvicorn.run(app, port=5173)
